libraries/matrix/impl/src/test/kotlin/io/element/android/libraries/matrix/impl/sdk/FakeRoomInterface.kt

```diff
@@ -0,0 +1,355 @@
+        TODO("Not yet implemented")
+    }
+
+    override fun enableSendQueue(enable: Boolean) {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun getPowerLevels(): RoomPowerLevels {
+        TODO("Not yet implemented")
+    }
+
+    override fun hasActiveRoomCall(): Boolean {
+        TODO("Not yet implemented")
+    }
+
+    override fun heroes(): List<RoomHero> {
+        TODO("Not yet implemented")
+    }
+
+    override fun id(): String {
+        return A_ROOM_ID.value
+    }
+
+    override suspend fun ignoreDeviceTrustAndResend(devices: Map<String, List<String>>, transactionId: String) {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun ignoreUser(userId: String) {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun inviteUserById(userId: String) {
+        TODO("Not yet implemented")
+    }
+
+    override fun invitedMembersCount(): ULong {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun inviter(): RoomMember? {
+        TODO("Not yet implemented")
+    }
+
+    override fun isDirect(): Boolean {
+        TODO("Not yet implemented")
+    }
+
+    override fun isEncrypted(): Boolean {
+        TODO("Not yet implemented")
+    }
+
+    override fun isPublic(): Boolean {
+        TODO("Not yet implemented")
+    }
+
+    override fun isSendQueueEnabled(): Boolean {
+        TODO("Not yet implemented")
+    }
+
+    override fun isSpace(): Boolean {
+        TODO("Not yet implemented")
+    }
+
+    override fun isTombstoned(): Boolean {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun join() {
+        TODO("Not yet implemented")
+    }
+
+    override fun joinedMembersCount(): ULong {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun kickUser(userId: String, reason: String?) {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun leave() {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun loadComposerDraft(): ComposerDraft? {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun markAsRead(receiptType: ReceiptType) {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun matrixToEventPermalink(eventId: String): String {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun matrixToPermalink(): String {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun member(userId: String): RoomMember {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun memberAvatarUrl(userId: String): String? {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun memberDisplayName(userId: String): String? {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun members(): RoomMembersIterator {
+        membersCallCount++
+        return getMembers()
+    }
+
+    override suspend fun membersNoSync(): RoomMembersIterator {
```

Comment:
@jmartinesp `members()` and `membersNoSync()` should return `RoomMembersIteratorInterface` instead, so that we can provide a Fake implementation.

Reply:
We *could* create a `RoomMembersInterface` trait in Rust, export that and make `RoomMembersIterator` implement it, but then on the FFI layer we'd have an interface created for the trait and then a class as the implementation:

```kotlin
public interface RoomMembersIteratorInterfaceInterface { // MOAR INTERFACES!
    
    fun `len`(): kotlin.UInt
    
    fun `nextChunk`(`chunkSize`: kotlin.UInt): List<RoomMember>?
    
    companion object
}

open class RoomMembersIteratorInterface: Disposable, AutoCloseable, RoomMembersIteratorInterfaceInterface {
...
```

And then:

```kotlin
// Same name as the open class above, but as an interface now?
public interface RoomMembersIteratorInterface {
    
    fun `len`(): kotlin.UInt
    
    fun `nextChunk`(`chunkSize`: kotlin.UInt): List<RoomMember>?
    
    companion object
}

// So is this using the interface above or the open class? I guess it's the interface
open class RoomMembersIterator: Disposable, AutoCloseable, RoomMembersIteratorInterface {
...
```

As the concrete Rust implementation ¯\\\_(ツ)_/¯. Does it make sense? Probably note. Is it useful for us in any way? Nope.

Some time ago I tried pushing a PR to the UniFFI project to ensure Rust functions took/returned the `*Interface` versions of the different exported structs and traits, but they said it might be too complex to maintain and cause some unexpected issues so I had to settle with just being able to create `open` versions of the concrete Rust classes, which is what we're currently using for `RoomMembersIterator` here.

Reply:
I do not understand, `RoomMembersIteratorInterface` already exists. But OK about about the limitation of the UniFFI.

Reply:
OK, let's extend open class then, since it seems to be the long term solution and it's not that bad.
If a method is not extended and used in the test, the test fails, to this is enough to have a not-so-bad devx.

https://github.com/element-hq/element-x-android/pull/3482/commits/eaaea7eb2c9cc9edb46777403c4b847de4cc4542

Reply:
`RoomMembersIteratorInterface` is an interface generated in the Kotlin bindings by UniFFI, not something that exists in Rust. It's supposed to be used for fakes, but every SDK call needs a concrete Rust type instead of the generated interface, so I don't really see how useful it can be. Apparently it does work well enough for the Mozilla folks, because they didn't want to change it 🫤 .